posthog/api/event.py

```diff
@@ -264,39 +264,36 @@ def get_values(self, request: request.Request) -> List[Dict[str, Any]]:
     # /event/sessions
     #
     # params:
-    # - offset: (number) offset query param for paginated list of user sessions
+    # - pagination: (dict) Object containing information about pagination (offset, last page info)
     # - distinct_id: (string) filter sessions by distinct id
     # - duration: (float) filter sessions by recording duration
     # - duration_operator: (string: lt, gt)
     # - **shared filter types
     # ******************************************
     @action(methods=["GET"], detail=False)
     def sessions(self, request: request.Request, *args: Any, **kwargs: Any) -> Response:
-        from posthog.queries.sessions_list import SESSIONS_LIST_DEFAULT_LIMIT, SessionsList
+        from posthog.queries.sessions.sessions_list import SessionsList
 
         filter = SessionsFilter(request=request)
-        limit = SESSIONS_LIST_DEFAULT_LIMIT + 1
-        result: Dict[str, Any] = {"result": SessionsList().run(filter=filter, team=self.team, limit=limit)}
+        pagination = json.loads(request.GET.get("pagination", "{}"))
 
-        if filter.distinct_id:
-            result = self._filter_sessions_by_distinct_id(filter.distinct_id, result)
+        sessions, pagination = SessionsList().run(filter=filter, team=self.team, **pagination)
 
-        if filter.session is None:
-            offset = filter.offset + limit - 1
-            if len(result["result"]) > SESSIONS_LIST_DEFAULT_LIMIT:
-                result["result"].pop()
-                date_from = result["result"][0]["start_time"].isoformat()
-                result.update({OFFSET: offset})
-                result.update({DATE_FROM: date_from})
+        if filter.distinct_id:
+            sessions = self._filter_sessions_by_distinct_id(filter.distinct_id, sessions)
 
-        return Response(result)
+        return Response({"result": sessions, "pagination": pagination})
 
-    def _filter_sessions_by_distinct_id(self, distinct_id: str, result: Dict[str, Any]) -> Dict[str, Any]:
+    def _filter_sessions_by_distinct_id(self, distinct_id: str, sessions: List[Any]) -> List[Any]:
         person_ids = Person.objects.get(persondistinctid__distinct_id=distinct_id).distinct_ids
-        result["result"] = [
-            session for i, session in enumerate(result["result"]) if result["result"][i]["distinct_id"] in person_ids
-        ]
-        return result
+        return [session for i, session in enumerate(sessions) if session["distinct_id"] in person_ids]
+
+    @action(methods=["GET"], detail=False)
+    def session_events(self, request: request.Request, *args: Any, **kwargs: Any) -> Response:
```

Comment:
does clickhouse also use this?

Reply:
There's a bit of feature drift here.

For clickhouse, the sessions query still returns events and the frontend knows to use those. For postgres, events are loaded separately (for perf reasons) -> separate query is needed.

Not sure what the correct approach here is, I don't think it's worth refactoring clickhouse query here right now. Thoughts?

Reply:
mm yeah I figured it was on these lines. I think if the frontend can handle then it's no problem. This will just be[ part of the consideration in consolidating](https://github.com/PostHog/ops/issues/131) our offering into clickhouse only. I think the duality will eventually be a major hinderance in moving quickly